java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -23,117 +23,210 @@ private typealias Path = List<Identifier>
+    constructor(id: Identifier, parent: Terminal) :
+      this(parent.accessPath + id, emptySet(), emptySet())
+
+    /** Adds influencing nodes to the `influence` edge set. */
+    override fun influencedBy(influence: Set<DependencyNode>) =
+      copy(influence = this.influence + influence.flatten())
+  }
+
+  /** A [Terminal] with input that has been modified by the expression. */
+  data class Derived(
+    override val accessPath: Path,
+    override val dependency: Set<Terminal>,
+    override val influence: Set<Terminal>
+  ) : Terminal() {
+    /** Expresses an access path with variable arguments.*/
+    constructor(
+      vararg paths: Identifier,
+      dependency: Set<Terminal> = emptySet(),
+      influence: Set<Terminal> = emptySet()
+    ) : this(listOf(*paths), dependency, influence)
+
+    /** Constructs an access path from a single [Identifier] and a known parent node. */
+    constructor(id: Identifier, parent: Terminal) :
+      this(parent.accessPath + id, emptySet(), emptySet())
+
+    /** Adds influencing nodes to the `influence` edge set. */
+    override fun influencedBy(influence: Set<DependencyNode>) =
+      copy(influence = this.influence + influence.flatten())
+  }
+
+  /** A collection of [Terminal] nodes. */
+  data class Nodes private constructor(
+    val nodes: Set<Terminal> = emptySet()
   ) : DependencyNode() {
+    constructor() : this(emptySet())
+    constructor(nodes: List<Terminal>) : this(nodes.toSet())
+
+    /** Converts [DependencyNode]s into a flattened collection of [Terminal] nodes. */
+    constructor(vararg nodes: DependencyNode) : this(listOf(*nodes).flatten())
 
-    /** Construct an [AssociationNode] from associations of [Identifier]s to [DependencyNode]s. */
-    constructor(vararg pairs: Pair<Identifier, DependencyNode>) : this(pairs.toMap())
+    /** Converts a list of node pairs into `dependency` relationships. */
+    constructor(vararg edges: Pair<Identifier, DependencyNode>) :
+      this(edges.groupBy({ it.first }, { it.second }).map { (id, node) ->
+        Equal(id, dependency = node.flatten())
+      })
 
-    /** Replace the associations of an [AssociationNode] with new mappings. */
-    fun add(vararg other: Pair<Identifier, DependencyNode>): DependencyNode = AssociationNode(
-      associations + other
+    /** Extends influence relationships to all contained nodes. */
+    override fun influencedBy(influence: Set<DependencyNode>) = copy(
+      nodes = nodes.map { it.influencedBy(influence) as Terminal }.toSet()
     )
+  }
 
-    /** Returns the [DependencyNode] associated with the input [Identifier]. */
-    fun lookup(key: Identifier): DependencyNode = requireNotNull(associations[key]) {
-      "Identifier '$key' is not found in AssociationNode."
-    }
+  /**
+   * Associates [Identifier]s with [DependencyNode]s and holds a buffer to express `influence`.
+   */
+  data class BufferedScope(
```

Comment:
I thought we decided to merge `Input` and `AssociatedNode` into one class. The only difference would have been that `Input` will have empty `associations` map?


Also, what is the difference between `AssociationNode` and `BufferedScope`?

Reply:
For most nodes (here, terminal nodes) and the future claim conversion, we can make do with just having a `path` (where a new node is defined as the parent node's path + the new id). This is done instead of needing to look up references in the associations map (recursively?).

Thus, `association` is only used for `Scopes`. In this state of the PR, I wanted to couple concerns together. Instead of using associations as proxy edges, I use the `dependency` field (see the `Nodes` constructor). But, there still needs to be a way to have an association of identifiers with nodes for adding to / reading from the scope (especially during lookups). BufferedScope serves this purpose, and it avoids us needed to make the visitor return a tuple of nodes: it acts like a tuple of scope and an influence buffer. Thus, it solves exactly the concern it's designed to address.